src/torch/README.md

```diff
@@ -0,0 +1,123 @@
+
+*   **Vector-Jacobian Product (VJP):** With `y = f(x)` as above, the VJP is the product `w^T J`, where `w` is a vector (often called the "upstream/puput gradient").  The VJP represents the gradient of a scalar-valued function that depends on `y` (e.g., a loss function), with respect to the input `x`.  *Reverse mode* autodiff computes VJPs. 
+
+*   **Hessian-Vector Product (HVP):** The Hessian matrix `H` contains all the second-order partial derivatives of a scalar-valued function `f(x)`. The HVP is the product `Hv`, where `v` is a vector.  It gives the second-order directional derivative of `f` at `x` in the direction of `v`. 
+
+## Wrapper Design
+
+The wrapper provides three main classes:
+
+1.  `SIRFTorchOperator`: Wraps a SIRF `Operator` (e.g., a projection operator). Computes the JVP forward and VJP in reverse.
+2.  `SIRFTorchObjectiveFunction`: Wraps a SIRF `ObjectiveFunction` for calculating its value forward, and gradient in reverse mode.
+3.  `SIRFTorchObjectiveFunctionGradient`: Wraps a SIRF `ObjectiveFunction` for calculating its gradient forward and computing the Hessian-vector product in reverse mode.
+
+These classes use custom `torch.autograd.Function` implementations (`_Operator`, `_ObjectiveFunction`, and `_ObjectiveFunctionGradient`) to define the forward and backward passes, handling the conversions between PyTorch tensors and SIRF objects.
+
+### `_Operator` (Forward and Backward Passes)
+
+*   **Forward Pass:**
+    1.  Converts the input PyTorch tensor to a SIRF object.
+    2.  Applies the SIRF `Operator.forward()` method.  Let's represent the SIRF Operator as `A`.  So, this step computes `y = Ax`, where `x` is the input (SIRF object) and `y` is the output (SIRF object).
+    3.  Converts the result back to a PyTorch tensor.
+    4.  If the input tensor requires gradients, it saves relevant information (the output SIRF object and the operator) in the context (`ctx`) for use in the backward pass.
+
+*   **Backward Pass (Adjoint Operation - VJP):**
+    1.  Receives the "upstream gradient" (`grad_output`) – the gradient of the loss `L` with respect to the *output* of the forward pass: `∂L/∂y`.
+    2.  Converts `grad_output` to a SIRF object.
+    3.  Applies the SIRF `Operator.backward()` method (which computes the adjoint operation). The adjoint operation, `A^T`, effectively computes the VJP.  If the forward pass computes `y = Ax`, the backward pass computes `x̄ = A^T ȳ`, where `A^T` is the adjoint of `A`, `ȳ` is `grad_output = ∂L/∂y`, and `x̄` is the gradient with respect to the input `x`. So, `x̄ = ∂L/∂x = (∂L/∂y)(∂y/∂x) =  ȳ^T J = (A^T)ȳ`. Here, `J` is the Jacobian of the forward operation `y=Ax`, and it is implicitly represented by the operator `A` itself.
```

Comment:
It is more general than the adjoint operation. I personally find it unhelpful to have these examples here. I would prefer to say `"Applies the SIRF `Operator.backward()` method on the output of step 2". That is really the only thing it does. Any other meanings depend on whatever `backward` happens to do (but this function doesn't know).

Reply:
I would argue that it's always the Jacobian adjoint, as it's always some first order approximation. I am greatly simplifying it but making sure I am consistent with the explanation... Let me know what you think once I have rewritten